More than one server running

I ran into this issue where if you have a server running already, you can't build the project again. The Gradle build will raise an error like:

    java.lang.AssertionError: Server did not start correctly. Perhaps it is already running?

Which is a useful error message, but would be better if handled so that it doesn't matter if you have a server running already.

--- a/cardshifter-server/src/test/java/com/cardshifter/server/main/ServerConnectionTest.java
+++ b/cardshifter-server/src/test/java/com/cardshifter/server/main/ServerConnectionTest.java
@@ -64,11 +64,14 @@ public class ServerConnectionTest {
 		PropertyConfigurator.configure(getClass().getResourceAsStream("log4j.properties"));
 		ServerConfiguration config = ServerConfiguration.defaults();
 
+		int basePortSocket = config.getPortSocket();
+		int basePortWebsocket = config.getPortWebsocket();
+
 		// The ports might be in use by another instance or application
 		// Could use port = 0, but would need access to the ServerSocket to get the real port number
 		for (int i = 0; i < MAX_SERVER_PORT_TRY; i++) {
-			config.setPortSocket(config.getPortSocket() + i * 10);
-			config.setPortWebsocket(config.getPortWebsocket() + i * 10);
+			config.setPortSocket(basePortSocket + i * 10);
+			config.setPortWebsocket(basePortWebsocket + i * 10);
 
 			main = new MainServer(config);
 			main.getMods().loadExternal(Paths.get("../extra-resources/groovy"));

--- a/cardshifter-server/src/test/java/com/cardshifter/server/main/ServerConnectionTest.java
+++ b/cardshifter-server/src/test/java/com/cardshifter/server/main/ServerConnectionTest.java
@@ -45,22 +45,45 @@ public class ServerConnectionTest {
 	private String getTestMod() {
         return mods.getMods()[0];
     }
+
+	private TestClient createTestClient() throws IOException {
+		return new TestClient(socketPort);
+	}
 	
 	private MainServer main;
 	private Server server;
+	private int socketPort;
 	private TestClient client1;
 	private int userId;
     private AvailableModsMessage mods;
 
+	private final int MAX_SERVER_PORT_TRY = 10;
+
     @Before
 	public void setup() throws IOException, InterruptedException {
 		PropertyConfigurator.configure(getClass().getResourceAsStream("log4j.properties"));
-		main = new MainServer(ServerConfiguration.defaults());
-        main.getMods().loadExternal(Paths.get("../extra-resources/groovy"));
-		server = main.start();
-		assertTrue("Server did not start correctly. Perhaps it is already running?", server.getClients().size() > 0);
-		
-		client1 = new TestClient();
+		ServerConfiguration config = ServerConfiguration.defaults();
+
+		// The ports might be in use by another instance or application
+		// Could use port = 0, but would need access to the ServerSocket to get the real port number
+		for (int i = 0; i < MAX_SERVER_PORT_TRY; i++) {
+			config.setPortSocket(config.getPortSocket() + i * 10);
+			config.setPortWebsocket(config.getPortWebsocket() + i * 10);
+
+			main = new MainServer(config);
+			main.getMods().loadExternal(Paths.get("../extra-resources/groovy"));
+			server = main.start();
+
+			if (server.getClients().size() > 0) {
+				break;
+			}
+		}
+
+		assertTrue("Server did not start correctly after " + MAX_SERVER_PORT_TRY + " retries.",
+				   server.getClients().size() > 0);
+
+		socketPort = config.getPortSocket();
+		client1 = createTestClient();
 		client1.send(new LoginMessage("Tester"));
 		
 		WelcomeMessage welcome = client1.await(WelcomeMessage.class);
@@ -84,8 +107,7 @@ public class ServerConnectionTest {
 	
 	@Test(timeout = 20000)
 	public void testUserOnlineOffline() throws InterruptedException, UnknownHostException, IOException {
-		
-		TestClient client2 = new TestClient();
+		TestClient client2 = createTestClient();
 		client2.send(new LoginMessage("Test2"));
 		client2.await(WelcomeMessage.class);
 		client2.await(ChatMessage.class);

--- a/cardshifter-server/src/test/java/com/cardshifter/server/main/TestClient.java
+++ b/cardshifter-server/src/test/java/com/cardshifter/server/main/TestClient.java
@@ -37,8 +37,8 @@ public class TestClient {
 	private final LinkedBlockingQueue<Message> messages = new LinkedBlockingQueue<>();
 	private final Thread thread;
 	
-	public TestClient() throws UnknownHostException, IOException {
-		this.socket = new Socket("127.0.0.1", 4242);
+	public TestClient(int port) throws UnknownHostException, IOException {
+		this.socket = new Socket("127.0.0.1", port);
 		this.mapper = CardshifterIO.mapper();
 		out = socket.getOutputStream();
 		in = socket.getInputStream();

--- a/cardshifter-server/src/test/java/com/cardshifter/server/messages/ExportTest.java
+++ b/cardshifter-server/src/test/java/com/cardshifter/server/messages/ExportTest.java
@@ -15,13 +15,37 @@ import com.cardshifter.server.main.ServerMain;
 import com.cardshifter.server.model.MainServer;
 import com.cardshifter.server.model.Server;
 
+import static org.junit.Assert.assertTrue;
+
 public class ExportTest {
+
+	private final int MAX_SERVER_PORT_TRY = 10;
 
 	@Test
 	public void test() {
 		PropertyConfigurator.configure(ServerMain.class.getResourceAsStream("log4j.properties"));
-		
-		Server server = new MainServer(ServerConfiguration.defaults()).start();
+
+		ServerConfiguration config = ServerConfiguration.defaults();
+		int basePortSocket = config.getPortSocket();
+		int basePortWebsocket = config.getPortWebsocket();
+		Server server = null;
+
+		// The ports might be in use by another instance or application
+		// Could use port = 0, but would need access to the ServerSocket to get the real port number
+		for (int i = 0; i < MAX_SERVER_PORT_TRY; i++) {
+			config.setPortSocket(basePortSocket + i * 10);
+			config.setPortWebsocket(basePortWebsocket + i * 10);
+
+			server = new MainServer(config).start();
+
+			if (server.getClients().size() > 0) {
+				break;
+			}
+		}
+
+		assertTrue("Server did not start correctly after " + MAX_SERVER_PORT_TRY + " retries.",
+				server != null && server.getClients().size() > 0);
+
 		String modName = server.getGameFactories().keySet().iterator().next();
 		ServerGame game = server.createGame(modName);
         FakeClient fakeClient = new FakeClient(server, e -> {});

